libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -230,15 +252,71 @@ internal object MapboxNavigationTelemetry :
 
     override fun onNavigationSessionStateChanged(navigationSession: NavigationSession.State) {
         log("session state is $navigationSession")
-        sessionState = navigationSession
         when (navigationSession) {
-            IDLE, FREE_DRIVE -> sessionStop()
+            IDLE, FREE_DRIVE -> {
+                sessionStop()
+                handleStateChanged(sessionState, navigationSession)
+            }
             ACTIVE_GUIDANCE -> {
                 locationsCollector.flushBuffers()
+                handleStateChanged(sessionState, navigationSession)
                 needStartSession = true
                 startSessionIfNeedAndCan()
             }
         }
+        sessionState = navigationSession
+    }
+
+    private fun handleStateChanged(
+        oldState: NavigationSession.State,
+        newState: NavigationSession.State
+    ) {
+        when {
+            oldState == FREE_DRIVE && newState == IDLE -> trackFreeDrive(STOP)
+            oldState == FREE_DRIVE && newState == ACTIVE_GUIDANCE -> trackFreeDrive(STOP)
+            oldState != FREE_DRIVE && newState == FREE_DRIVE -> trackFreeDrive(START)
+        }
+    }
+
+    private fun trackFreeDrive(type: FreeDriveEventType) {
+        log("trackFreeDrive $type")
+
+        fun createFreeDriveEvent() {
+            log("createFreeDriveEvent")
+            if (type == START) {
+                dynamicFreeDriveValues.run {
+                    sessionId = obtainUniversalUniqueIdentifier()
+                    sessionStartTime = Date()
+                }
+            }
+
+            val freeDriveEvent = NavigationFreeDriveEvent(PhoneState(context)).apply {
+                populate(type)
+            }
+
+            if (type == STOP) {
+                dynamicFreeDriveValues.reset()
+            }
+
+            log("${freeDriveEvent::class.java} event sent")
+            metricsReporter.addEvent(freeDriveEvent)
+        }
+
+        if (locationsCollector.lastLocation == null) {
+            log("wait for location")
+            jobControl.scope.launch {
+                var waitForLocation = true
+                while (waitForLocation) {
+                    delay(50)
```

Comment:
no, can't avoid it. 
`locationsCollector.lastLocation` is a `Location` , so we can't check it with a callback or any other mechanism.
`delay` doesn't have any cons, it doesn't block. Just asks for `lastLocation` until it's non null

Reply:
Iirc @kmadsen implemented this without `delay` using coroutines? Is that right? Could you confirm @kmadsen?

Reply:
@Guardiola31337 what concerns do you have related to `delay`?

Reply:
I was searching for an elegant pattern or framework here because rxjava has a nice solution. I prefer this type of logic inside frameworks that are broadly used.  

That said, I also did the same thing as @korshaknn. Here's the function from the driver app and it is using `delay`. Also the replay [simulator is using delay](https://github.com/mapbox/mapbox-navigation-android/blob/740b47fb50daeaf25cc413773241e1e29a0c905a/libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayEventSimulator.kt#L38) and it works great.
``` kotlin
    private suspend fun firstLocation(): Location {
        var nonNullLocation: Location? = location
        while (nonNullLocation == null) {
            delay(100L)
            nonNullLocation = location
        }
        return nonNullLocation
    }
```

That said, the new hotness is kotlin `Flow`. If I were to take another pass at a solution for this, I would try to use `MutableStateFlow`. Something like this.

``` kotlin
    val currentLocationFlow = MutableStateFlow<Location?>(null)
    private suspend fun firstLocation(callback: (Location) -> Unit) {
        currentLocationFlow.filterNotNull()
            .collect { location ->
                callback(location)
            }
    }
```

Reply:
Thanks for confirming @kmadsen